optuna/_callbacks.py

```diff
@@ -55,3 +55,58 @@ def __call__(self, study: "optuna.study.Study", trial: FrozenTrial) -> None:
         n_complete = len(trials)
         if n_complete >= self._n_trials:
             study.stop()
+
+
+@experimental("2.8.0")
+class RetryFailedTrialCallback:
+    """Retry a failed trial up to a maximum number of times.
+
+    When a trial fails, this callback can be used with the :class:`optuna.storage` class to
+    recreate the trial in :obj:`TrialState.WAITING` to queue up the trial to be run again.
+
+    This is helpful in environments where trials may fail due to external conditions, such as
+    being pre-empted by other processes.
+
+    Usage:
+
+        .. testcode::
+
+            import optuna
+            from optuna.storages import RetryFailedTrialCallback
+
+            storage = optuna.storages.RDBStorage(
+                url="sqlite:///:memory:",
+                heartbeat_interval=60,
+                grace_period=120,
+                failed_trial_callback=RetryFailedTrialCallback(max_retry=3),
+            )
+
+            study = optuna.create_study(
+                storage=storage,
+            )
+
+    """
+
+    def __init__(self, max_retry: int) -> None:
+        self._max_retry = max_retry
+
+    def __call__(self, study: "optuna.study.Study", trial: FrozenTrial) -> None:
+        retries = sum(
+            ("failed_trial", trial.number) in s.system_attrs.items() for s in study.trials
+        )
+
+        if retries + 1 > self._max_retry:
+            return
+
+        system_attrs = {"failed_trial": trial.number}
+        system_attrs.update(trial.system_attrs)
```

Comment:
Is the expectation that this line should update/overwrite the value of the `"failed_trial"` key with the very first trial number if this is not the first retry? Might be good to leave a comment since the logic isn't obvious.

Reply:
Looks great, thanks!